x-pack/filebeat/input/awss3/s3_filters.go

```diff
@@ -0,0 +1,124 @@
+// you may not use this file except in compliance with the Elastic License.
+
+package awss3
+
+import (
+	"sync"
+	"time"
+
+	"github.com/elastic/elastic-agent-libs/logp"
+)
+
+const (
+	filterOldestTime = "oldestTimeFilter"
+	filterStartTime  = "startTimeFilter"
+)
+
+// filterProvider exposes filters that needs to be applied on derived state.
+// Once configured, retrieve filter applier using getApplierFunc.
+type filterProvider struct {
+	cfg *config
+
+	staticFilters []filter
+	once          sync.Once
+}
+
+func newFilterProvider(cfg *config) *filterProvider {
+	fp := &filterProvider{
+		cfg: cfg,
+	}
+
+	// derive static filters
+	if cfg.StartTimestamp != "" {
+		// note - errors should not occur as this has validated prior reaching here
+		parse, _ := time.Parse(time.RFC3339, cfg.StartTimestamp)
+		fp.staticFilters = append(fp.staticFilters, newStartTimestampFilter(parse))
+	}
+
+	return fp
+}
+
+// getApplierFunc returns aggregated filters valid at the time of retrival.
+// Applier return true if state is valid for processing according to the underlying filter collection.
+func (f *filterProvider) getApplierFunc() func(log *logp.Logger, s state) bool {
+	filters := map[string]filter{}
+
+	if f.cfg.IgnoreOlder != 0 {
+		timeFilter := newOldestTimeFilter(f.cfg.IgnoreOlder)
+		filters[timeFilter.getID()] = timeFilter
+	}
+
+	for _, f := range f.staticFilters {
+		filters[f.getID()] = f
+	}
+
+	f.once.Do(func() {
+		// Ignore the oldest time filter once for initial startup only if start timestamp filter is defined
+		// This allows users to ingest desired data from start time onwards.
+		if filters[filterStartTime] != nil {
+			delete(filters, filterOldestTime)
+		}
+	})
+
+	return func(log *logp.Logger, s state) bool {
+		for _, f := range filters {
+			if !f.isValid(s) {
+				log.Debugf("skipping processing of object '%s' by filter '%s'", s.Key, f.getID())
+				return false
+			}
+		}
+
+		return true
+	}
+}
+
+// filter defines the fileter implementation contract
+type filter interface {
+	getID() string
+	isValid(objState state) (valid bool)
+}
+
+// startTimestampFilter - filter out entries based on provided start time stamp, compared to filtering object's last modified times stamp.
+type startTimestampFilter struct {
+	id        string
+	timeStart time.Time
+}
+
+func newStartTimestampFilter(start time.Time) *startTimestampFilter {
+	return &startTimestampFilter{
+		id:        filterStartTime,
+		timeStart: start,
+	}
+}
+
+func (s startTimestampFilter) isValid(objState state) bool {
+	return s.timeStart.Before(objState.LastModified)
+}
+
+func (s startTimestampFilter) getID() string {
+	return s.id
+}
+
+// oldestTimeFilter - filter out entries based on calculated oldest modified time, compared to filtering object's last modified times stamp.
+type oldestTimeFilter struct {
+	id         string
+	timeOldest time.Time
+}
+
+func newOldestTimeFilter(timespan time.Duration) *oldestTimeFilter {
+	oldest := time.Now().Add(-1 * timespan)
```

Comment:
I think it would be easier if the signature would be:

```go
// NewOldestTimeFilter creates a new oldestTimeFilter based on the given timespan.
func NewOldestTimeFilter(timespan time.Duration, now time.Time) *oldestTimeFilter {
	oldest := now.Add(-1 * timespan)
```

Instead of calculating the time there. It would make it easier to test this too.

Also, the functions from this struct don't use pointer, is that intentional?

Reply:
I adapted the proposal on the method signature. And yes, it helps with testing :) 

> Also, the functions from this struct don't use pointer, is that intentional?

yes, this is intentional as there's no change in the derived struct throughout its usage